os/kernel/binary_manager/compression/compress_read.c

```diff
@@ -0,0 +1,409 @@
+		if (size < 0) {
+			bmdbg("Read for compressed block %d failed\n", index);
+			buffer_index = size;
+			goto error_compress_read;
+		}
+
+		/* Decompress block in read_buffer to out_buffer */
+		ret = compress_decompress_block(buffers.out_buffer, &writesize, buffers.read_buffer, &size, index);
+		if (ret == ERROR) {
+			bmdbg("Failed to decompress %d block of this binary\n", index);
+			buffer_index = ret;
+			goto error_compress_read;
+		}
+
+		if (index == first_block) {
+			/*
+			 * Read appropriate number of bytes to buffer from this block using readsize and offset info.
+			 * If start_offset + readsize < end_offset, write from start_offset to start_offset + readsize.
+			 * Otherwise, write from start_offset to end_offset into buffer.
+			 */
+			block_size_to_write = ((index + 1) * blocksize - 1 > actual_offset + readsize - 1 ? readsize : (index + 1) * blocksize - actual_offset);
+			memcpy(&buffer[buffer_index], &buffers.out_buffer[actual_offset - (index * blocksize)], block_size_to_write);
+			buffer_index += block_size_to_write;
+		} else if (index == last_block) {
+			/*
+			 * Read appropriate number of bytes to buffer from this block using readsize and offset info.
+			 * Calculate end_offset (Offset of last byte to write from this block).
+			 * Write from start_offset to end_offset from this block into buffer.
+			 */
+			block_size_to_write = actual_offset + readsize - (index * blocksize);
+			memcpy(&buffer[buffer_index], &buffers.out_buffer[0], block_size_to_write);
+			buffer_index += block_size_to_write;
+		} else {
+			/*
+			 * This is not the first or last block. This is intermediary block.
+			 * So, write entire block into buffer.
+			 */
+			block_size_to_write = blocksize;
+			memcpy(&buffer[buffer_index], &buffers.out_buffer[0], block_size_to_write);
+			buffer_index += block_size_to_write;
+		}
+	}
+
+error_compress_read:
+	return buffer_index;
+}
+
+/****************************************************************************
+ * Name: compress_init
+ *
+ * Description:
+ *   Initialize the compression_header of type'struct s_header' for this file
+ *
+ * Returned value:
+ *   OK (0) on Success
+ *   ERROR (-1) on Failure
+ ****************************************************************************/
+int compress_init(int filfd, uint16_t offset, off_t *filelen)
+{
+	int ret;
+
+	/* Parsing compression header for compressed file */
+	ret = compress_parse_header(filfd, offset);
+	if (ret != OK) {
+		bmdbg("Failed to parse compression header from file\n");
+		goto error_compress_init;
+	}
+
+	/* Assign file length as that of uncompressed file */
+	*filelen = compression_header.binary_size;
+
+#if CONFIG_COMPRESSION_TYPE == 1
+	/* Allocating memory for read and out buffer to be used for LZMA decompression */
+	if (compression_header.compression_format == LZMA) {
+		buffers.read_buffer = (unsigned char *)malloc(compression_header.blocksize + 5);
```

Comment:
Please check whether malloc is successful or not.

Reply:
Tested on device, it is working.